src/common/src/array/list_array.rs

```diff
@@ -431,118 +376,189 @@ impl ListValue {
+        Self::new(iter.into_iter().collect::<BoolArray>().into())
+    }
+}
+
+impl<'a> FromIterator<Option<&'a str>> for ListValue {
+    fn from_iter<I: IntoIterator<Item = Option<&'a str>>>(iter: I) -> Self {
+        Self::new(iter.into_iter().collect::<Utf8Array>().into())
+    }
+}
+
+impl<'a> FromIterator<&'a str> for ListValue {
+    fn from_iter<I: IntoIterator<Item = &'a str>>(iter: I) -> Self {
+        Self::new(iter.into_iter().collect::<Utf8Array>().into())
+    }
+}
+
+impl FromIterator<ListValue> for ListValue {
+    fn from_iter<I: IntoIterator<Item = ListValue>>(iter: I) -> Self {
+        Self::new(iter.into_iter().collect::<ListArray>().into())
+    }
+}
+
+impl From<ListValue> for ArrayImpl {
+    fn from(value: ListValue) -> Self {
+        *value.values
     }
 }
 
 #[derive(Copy, Clone)]
-pub enum ListRef<'a> {
-    Indexed { arr: &'a ListArray, idx: usize },
-    ValueRef { val: &'a ListValue },
+pub struct ListRef<'a> {
+    array: &'a ArrayImpl,
+    start: u32,
+    end: u32,
 }
 
 impl<'a> ListRef<'a> {
     /// Returns the length of the list.
     pub fn len(&self) -> usize {
-        match self {
-            ListRef::Indexed { arr, idx } => (arr.offsets[*idx + 1] - arr.offsets[*idx]) as usize,
-            ListRef::ValueRef { val } => val.values.len(),
-        }
+        (self.end - self.start) as usize
     }
 
     /// Returns `true` if the list has a length of 0.
     pub fn is_empty(&self) -> bool {
-        self.len() == 0
+        self.start == self.end
     }
 
-    /// Returns the elements in the flattened list.
-    pub fn flatten(self) -> Vec<DatumRef<'a>> {
-        // XXX: avoid using vector
-        iter_elems_ref!(self, it, {
-            it.flat_map(|datum_ref| {
-                if let Some(ScalarRefImpl::List(list_ref)) = datum_ref {
-                    list_ref.flatten()
-                } else {
-                    vec![datum_ref]
-                }
-                .into_iter()
-            })
-            .collect()
-        })
+    /// Returns the data type of the elements in the list.
+    pub fn data_type(&self) -> DataType {
+        self.array.data_type()
     }
 
-    /// Returns the total number of elements in the flattened list.
-    pub fn flatten_len(self) -> usize {
-        iter_elems_ref!(self, it, {
-            it.map(|datum_ref| {
-                if let Some(ScalarRefImpl::List(list_ref)) = datum_ref {
-                    list_ref.flatten_len()
-                } else {
-                    1
-                }
-            })
-            .sum()
-        })
+    /// Returns the elements in the flattened list.
+    pub fn flatten(self) -> ListRef<'a> {
+        match self.array {
+            ArrayImpl::List(inner) => ListRef {
+                array: &inner.value,
+                start: inner.offsets[self.start as usize],
+                end: inner.offsets[self.end as usize],
+            }
+            .flatten(),
```

Comment:
So amazing. 🤩 So we're actually recursively mapping with the `offsets`?

Reply:
Yes. 🤯

P.S. Recommend a game called [Cocoon](https://store.steampowered.com/app/1497440/COCOON/). It helps you upgrade your brain on recursion.

Reply:
Yeah! I already played that on Xbox. It‘s really brain-storming when self-referencing appears when the game is close to the end.